plasTeX/ConfigManager.py

```diff
@@ -0,0 +1,181 @@
+
+    def setFromString(self, string: str):
+        self.value.extend(shlex.split(string))
+
+    def updateFromDict(self, data: Dict["str", Any]):
+        value = data.get(self.name)
+        if value is not None:
+            for entry in value:
+                self._value.extend(entry)
+
+class DictOption(ABC, Option[Dict[str, T]]):
+    @classmethod
+    @abstractmethod
+    def entryFromString(cls, entry: str) -> T:
+        """This convers a string into T"""
+
+    def set(self, key: str, value: str):
+        self.value[key] = self.entryFromString(value)
+
+    def setFromString(self, string: str):
+        raise ValueError("Calling setFromString on a DictOption. The individual items should be inserted with set.")
+
+    def updateFromDict(self, data: Dict["str", Any]):
+        entries = data.get(self.name)
+        if entries is not None:
+            for key, val in entries:
+                self.set(key, val)
+
+class ConfigSection():
+    def __init__(self, name: str, parent: 'ConfigManager'):
+        self.name = name
+        self.parent = parent
+        self.data = {} # type: Dict[str, Option]
+
+    def __getitem__(self, key: str) -> Any:
+        value = self.data[key].value
+        wrapper = InterpolationWrapper(self.parent)
+        if isinstance(value, str):
+            return value % wrapper
+        elif isinstance(value, list) and len(value) > 0 and isinstance(value[0], str):
+            return [x % wrapper for x in value]
+        else:
+            return value
+
+    def __setitem__(self, key: str, value: Any):
+        if isinstance(value, Option):
+            if key in self.data:
+                raise ValueError
+            self.data[key] = value
+        else:
+            self.data[key].value = value
+
+    def get(self, key: str, default: Optional[Any]=None) -> Optional[Any]:
+        try:
+            return self[key]
+        except KeyError:
+            return default
+
+    def keys(self) -> KeysView[str]:
+        return self.data.keys()
+
+    def registerArgparse(self, parser: ArgumentParser):
+        group = parser.add_argument_group(self.name)
+
+        for option in self.data.values():
+            option.registerArgparse(group)
+
+    def updateFromDict(self, data: Dict[str, Any]):
+        for option in self.data.values():
+            option.updateFromDict(data)
+
+class ConfigManager(Dict[str, ConfigSection]):
+    def registerArgparse(self, parser: ArgumentParser):
+        for item in self.values():
+            item.registerArgparse(parser)
+
+    def updateFromDict(self, data: Dict[str, Any]):
+        for section in self.values():
+            section.updateFromDict(data)
+
+    def addSection(self, name: str, description: Optional[str]=None) -> ConfigSection:
+        if name in self:
+            raise ValueError("Section {} already exists".format(name))
+
+        if description is None:
+            description = name.capitalize() + " Options"
```

Comment:
```suggestion
        description = description or name.capitalize() + " Options"
```

Reply:
This behaves differently if `description = ""`.

Reply:
Yes, I understand that. But don't you want the default description in that case?

Reply:
I'd imagine if someone explicitly set description to "", that's what they want it to be.